1-theory/1-apis-services-microservices.adoc

```diff
@@ -0,0 +1,144 @@
+
+Even if compliance isn´t an issue, comminicating over the network through standardized protocals make it possible for different
+services to be written in different programming languages or different frameworks. Instead of using the same language for everything, you 
+can use the right tool for the job, selecting a programming language that has strengths for certain tasks.
+
+WARNING: One thing to be wary of here is that some developers get a little giddy when told they can "use whatever language you like", finally giving CoolNewLanguage™ a try, even though the are not
+very experienced with it. Not only will that make it hard to maintain for that developer, it might also make it hard to find other developers
+who can help maintain it. If management are not careful, a company just getting into SoA can have 20 services in 20 different languages, and 
+this increases the danger of "Bus Factor" more than my 8 mile bike commute in NYC. 
+
+The positives are fairly clear, but an often overlooked problem with service-orientated arcitecture is the complexity of managing multiple services, and not just the comlpications of handling "more servers". Managing two services is not twice as hard as managing a single application, it is infinitely more complex, and you need people who know how to handle these issues. 
+
+In a monolith, if you try and call a class, unless your tests were garbage and that class didnt exist, there is zero problem with calling it. With services, they are going over the network to another server. The server might have changed its IP or domain name. It could be out of capacity, or have crashed entirely. Maybe the team who manages it deployed some changes and didn´t tell you, and the test suite for your API is not aware of their changes either... 
+
+The pros usually outweigh the cons so long as you have good devops culture, or dedicated sysops people, but SoA is not a merry world of unicorns as some would have you believe.
+
+Anyway, these applications need to interact with each other, and their clients need to interact with them, otherwise they 
+probably arent really doing anything. The interaction could happen in a bunch of different ways over the network, 
+but again that is usually over HTTP, AMQP, or some other messaging protocol. The interaction is by definition a form of API. 
+
+So, an API and a service are not different things. A service will always have some form of API, but not all APIs are a service...
+
+// TODO figure out where to put this, got it approved https://twitter.com/dhelper/status/1101140486130159617?s=12
+
+== Microservice
+
+
+Microservices as a term was coined to be a meaningful off-shoot of SoA, and has has been floating around more and more since 2011. As an off-shoot, many of the same pros and cons of SoA apply, with a few more pros and cons to mention on top.
```

Comment:
```suggestion
Microservices as a term was coined to be a meaningful off-shoot of SoA, and has been floating around more and more since 2011. As an off-shoot, many of the same pros and cons of SoA apply, with a few more pros and cons to mention on top.
```